### ISSUE TITLE
does it only keras or backend function can be used in custom loss function?

### ISSUE THREAD

**Issue by wl3b10s** [keras-team/keras#12057]
when defining a custom loss function, i want to:
1) operate the input y_pred tensor with numpy 
2) process  the input y_pred tensor to a external binary execution and get a result

demo code like below:

def mse_my_loss(y_true,y_pred):

    Q1 = K.mean(K.square(y_pred - y_true), axis=-1)    
    y_in = y_pred.numpy()	
    file_wave = './tools/data.raw'
    librosa.output.write_wav(file_wave, y_pred, 16000)
    CMD = './tools/EVAL +16000 ' + file_wave
    os.system(CMD)    
    loss = readresult()	
    return Q1 + loss  

whether it's feasible to achieve this?

Or only keras or backend function can be used in a customed loss function to let keras perform automatic gradient calculation on tensor？

thanks a lot.

**Comment by ParikhKadam** [comment]
@wl3b10s 

Try this.. your code won't get executed. Tensorflow tries to make a path between the input and output of the model. If any tensor is calculated at a point but is not actually needed to reach the destination, it will be ignored at runtime (training/testing).

The same applies for loss functions. Tensorflow will try to create a path from y_true and y_pred to the return value (here, Q1 + loss). If other operations are not important or are not in path, then they will be ignored.

If you need other operations to be performed, you need to mention it in the path from inputs to the returning value; that is, your output should be dependent on that operations and you **must** mention it in your code.

Yes.. you can't use numpy operations in Keras. Use backend or Keras itself.

I would like to stay updated on this.. Maybe I am wrong at a place but this is what I learned with trial and error. Thank you..

**Comment by wl3b10s** [comment]
thanks for  your help. @ParikhKadam .

**Comment by ParikhKadam** [comment]
Welcome @wl3b10s

**Comment by msymp** [comment]
This issue is closed. Thanks @ParikhKadam .

**Comment by ParikhKadam** [comment]
@msymp Welcome..
